packages/ra-core/src/controller/list/WithListContext.tsx

```diff
@@ -20,13 +20,40 @@ import { useListContext } from './useListContext';
  * );
  */
 export const WithListContext = <RecordType extends RaRecord>({
+    empty,
+    loading,
+    errorElement,
     render,
-}: WithListContextProps<RecordType>) =>
-    render(useListContext<RecordType>()) || null;
+    ...props
+}: WithListContextProps<RecordType>) => {
+    const context = useListContextWithProps<RecordType>(props);
+    const { data, total, isPending, error } = context;
 
-export interface WithListContextProps<RecordType extends RaRecord> {
+    if (isPending === true && loading) {
```

Comment:
loading=null is a valid ReactNode and should be taken into account. I thing you should change the test to be as in https://github.com/marmelab/react-admin/pull/10864/files#diff-cdac9daf5ac7f12fe651a4357b777c27564f7372b0387256dd630f0a9d07c479